server/src/lib/model/achievements.ts

```diff
@@ -0,0 +1,128 @@
+import { ChallengeToken, AchievementType } from 'common/challenge';
+import { getMySQLInstance } from './db/mysql';
+
+const bonus_condition_sql = {
+  // Arguments: [challenge, client_id]
+  // No need to check if user has already won the bonus because it must happend after a successful enrollment.
+  sign_up_first_three_days: `
+    SELECT enrolled_at < TIMESTAMPADD(DAY, 3, challenges.start_date) AS win_bonus, client_id AS bonus_winner
+    FROM enroll
+    LEFT JOIN challenges ON enroll.challenge_id = challenges.id AND challenges.url_token = ?
+    WHERE client_id = ?
+    `,
+  // Arguments: [client_id, invite, invite, challenge]
+  // Make sure invitees are enrolled during the challenge, no need to do that for the invitor
+  invite_signup: `
+    SELECT (invitee.id IS NOT NULL) AS win_bonus, invitor.client_id AS bonus_winner
+    FROM challenges
+    LEFT JOIN enroll invitee ON invitee.client_id = ?
+        AND invitee.invited_by = ? AND invitee.enrolled_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
+    LEFT JOIN enroll invitor ON invitor.url_token = ?
+    WHERE challenges.url_token = ?
+    `,
+  // Arguments: [client_id, client_id, challenge]
+  // To earn an three_day_streak bonus, user has to satisfy all the following conditions:
+  // 1) has been enrolled;
+  // 2) not earned three_day_streak bonus before;
+  // 3) already set a customer goal
+  // 4) already reached a 3-day-streak goal OR haven't reached such a goal but has streak activity that lasts for AT LEAST 3 days
+  //    just in case no such bonus are missed
+  three_day_streak: `
+    SELECT
+        (enroll.client_id IS NOT NULL) AND (earn.client_id IS NULL) AND (custom_goals.id IS NOT NULL) AND ((reached_goals.client_id IS NOT NULL) OR (streaks.id IS NOT NULL)) AS win_bonus,
+        ? AS bonus_winner
+    FROM challenges
+    LEFT JOIN achievements ON achievements.name = 'three_day_streak' AND challenges.id = achievements.challenge_id
+    LEFT JOIN enroll ON enroll.client_id = ?
+        AND challenges.id = enroll.challenge_id
+        AND enroll.enrolled_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
```

Comment:
We have a bunch of these checks littered throughout the code, which makes me think it should be extracted from our DB into a more reusable utility. If it's outside of the start/end date, a user shouldn't be able to enroll in the first place.

Reply:
Yes, it totally make sense. Actually I also thought about it, just not had the big picture before.